verilog/CST/class_test.cc

```diff
@@ -0,0 +1,136 @@
+#include "common/util/logging.h"
+#include "common/util/range.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "verilog/analysis/verilog_analyzer.h"
+
+#undef ASSERT_OK
+#define ASSERT_OK(value) ASSERT_TRUE((value).ok())
+
+namespace verilog {
+namespace {
+
+using verible::down_cast;
+using verible::SyntaxTreeNode;
+using verible::SyntaxTreeSearchTestCase;
+using verible::TokenInfoTestData;
+using verible::TreeSearchMatch;
+
+TEST(GetClassNameTest, ClassName) {
+  constexpr int kTag = 1;  // value doesn't matter
+  const SyntaxTreeSearchTestCase kTestCases[] = {
+      {""},
+      {"module m(); endmodule: m"},
+      {"class ", {kTag, "foo"}, ";\nendclass"},
+      {"class ",
+       {kTag, "foo"},
+       ";\nendclass\n class ",
+       {kTag, "bar"},
+       ";\n endclass"},
+      {"module m();\n class ", {kTag, "foo"}, ";\n endclass\n endmodule: m\n"},
+      {"class ",
+       {kTag, "foo"},
+       ";\nclass ",
+       {kTag, "bar"},
+       "; endclass\nendclass"},
+  };
+  for (const auto& test : kTestCases) {
+    const absl::string_view code(test.code);
+    VerilogAnalyzer analyzer(code, "test-file");
+    const auto code_copy = analyzer.Data().Contents();
+    ASSERT_OK(analyzer.Analyze()) << "failed on:\n" << code;
+    const auto& root = analyzer.Data().SyntaxTree();
+
+    const auto decls = FindAllClassDeclarations(*ABSL_DIE_IF_NULL(root));
+
+    std::vector<TreeSearchMatch> names;
+    for (const auto& decl : decls) {
+      const auto& type = GetClassName(*decl.match);
+      names.push_back(TreeSearchMatch{&type, {/* ignored context */}});
+    }
+
+    std::ostringstream diffs;
+    EXPECT_TRUE(test.ExactMatchFindings(names, code_copy, &diffs))
+        << "failed on:\n"
+        << code << "\ndiffs:\n"
+        << diffs.str();
+  }
+}
+
+TEST(GetClassNameTest, ClassEndLabel) {
+  constexpr int kTag = 1;  // value doesn't matter
+  const SyntaxTreeSearchTestCase kTestCases[] = {
+      {""},
+      {"module m(); endmodule: m"},
+      {"class foo;\nendclass: ", {kTag, "foo"}},
+      {"class foo;\nendclass: ",
+       {kTag, "foo"},
+       "\n class bar;\n endclass: ",
+       {kTag, "bar"}},
+      {"module m();\n class foo;\n endclass: ",
+       {kTag, "foo"},
+       "\n endmodule: m\n"},
+      {"class foo;\nclass bar;\n endclass: ",
+       {kTag, "bar"},
+       "\nendclass: ",
+       {kTag, "foo"}},
+  };
+  for (const auto& test : kTestCases) {
+    const absl::string_view code(test.code);
+    VerilogAnalyzer analyzer(code, "test-file");
+    const auto code_copy = analyzer.Data().Contents();
+    ASSERT_OK(analyzer.Analyze()) << "failed on:\n" << code;
+    const auto& root = analyzer.Data().SyntaxTree();
+
+    const auto decls = FindAllClassDeclarations(*ABSL_DIE_IF_NULL(root));
+
+    std::vector<TreeSearchMatch> types;
+    for (const auto& decl : decls) {
+      const auto* type = GetClassEndLabel(*decl.match);
+      types.push_back(TreeSearchMatch{type, {/* ignored context */}});
+    }
```

Comment:
Food for future thought: In many of the tests you've seen and written, this section of code is the only block that changes, while everything else remains constant boilerplate.  Could the testing functions be refactored to accept a lambda to reduce the amount of boilerplate copying?  O_o

No change needed for this PR.